ES|QL Add initial support for semantic_text field type

ioanatia opened the pull request:
Support is added behind a feature flag. We could not simply use `EsqlCapabilities` since that's not available in `esql-core`.
Right now we have no support in existing functions.
I followed the approach for adding initial support for `date_nanos` which was also added behind a feature flag which allowed for incremental progress, rather than adding support for everything in one PR: https://github.com/elastic/elasticsearch/pull/110205

With this PR we will return `semantic_text` fields as part of the results and it will also allow us to refer to `semantic_text` fields in the `match` function (to run semantic search):

```
{
  "columns": [
    {
      "name": "semantic_text_field",
      "type": "semantic_text"
    }
  ],
  "values": [
    [
      "these are not the droids you're looking for. He's free to go around"
    ],
    [
      "all we have to decide is what to do with the time that is given to us"
    ],
    [
      "I have no idea what I'm doing, but I know I'm doing it really, really well."
    ]
  ]
}
```

realtes: #115103

ioanatia committed:
Add initial support for semantic_text field type

ioanatia committed:
Update docs/changelog/113920.yaml

elasticsearchmachine commented:
Hi @ioanatia, I've created a changelog YAML for you.

ioanatia committed:
More tests and fixes

ioanatia committed:
Use mock inference service

ioanatia committed:
Fix tests

ioanatia committed:
Spotless

ioanatia committed:
Fix mixed-cluster and multi-clusters tests

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

ioanatia commented on x-pack/plugin/esql/qa/server/multi-clusters/src/javaRestTest/java/org/elasticsearch/xpack/esql/ccq/Clusters.java:
```diff
@@ -43,6 +44,7 @@ public static ElasticsearchCluster localCluster(ElasticsearchCluster remoteClust
             .setting("cluster.remote.connections_per_cluster", "1")
             .shared(true)
             .setting("cluster.routing.rebalance.enable", "none")
+            .plugin("inference-service-test")
```
this plugin is used for testing `semantic_text` - it's an inference service that can create sparse or dense embeddings - that have no actual "semantic" meaning since they are not using a model, but they are supposed to be deterministic.

ioanatia commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:
```diff
@@ -236,15 +242,45 @@ public static void loadDataSetIntoEs(RestClient client, Logger logger) throws IO
+        Request request = new Request("PUT", "_inference/sparse_embedding/test_sparse_inference");
+        request.setJsonEntity("""
+                  {
+                   "service": "test_service",
```
this is the inference service from the `inference-service-test` test plugin

nik9000 replied:
This one's probably worth javadoc to explain that it's for the semantic text fields.

ChrisHegarty replied:
It kinda surprises me that this is the first capability conditional in the test infra.  How expensive is this service endpoint? Conditionally creating it is of course fine, but the test infra would be simpler if it was unconditionally registered.

carlosdelest replied:
The test endpoint is quite lightweight. We could always register it.

ioanatia replied:
javadoc was added

nik9000 commented on x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/EsqlSpecIT.java:
```diff
@@ -14,7 +14,7 @@
 public class EsqlSpecIT extends EsqlSpecTestCase {
     @ClassRule
-    public static ElasticsearchCluster cluster = Clusters.testCluster(spec -> {});
+    public static ElasticsearchCluster cluster = Clusters.testCluster(spec -> spec.plugin("inference-service-test"));
```
Will everything need this plugin? should it be in `testCluster`?

ioanatia replied:
I think only this one needs it - it's the only one for `multi_node` that extends from `EsqlSpecTestCase` that loads the CSV data sets.

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec:
```diff
@@ -0,0 +1,10 @@
+simple
+required_capability: semantic_text_type
+
+FROM semantic_text;
```
You'll have to `| SORT` this or else it'll be jumbled in any multi-node tests.

ChrisHegarty replied:
good catch. done.

nik9000 commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:
```diff
@@ -571,6 +573,12 @@ private String generateInvalidQueryInferenceResultsMessage(StringBuilder baseMes
+        @Override
+        public BlockLoader blockLoader(MappedFieldType.BlockLoaderContext blContext) {
+            SourceValueFetcher fetcher = SourceValueFetcher.toString(blContext.sourcePaths(name().concat(".text")));
+            return new BlockSourceReader.BytesRefsBlockLoader(fetcher, BlockSourceReader.lookupMatchingAll());
```
Someone else'll have to verify that this is the right way to load these. It's how you load text, after all. But if there's a nicer docvalues way this is stored we should use that if we can.

tteofili replied:
yeah, I think this is somewhat dependent on how `semantic_text` fields store text in ES/Lucene. 

In a ideal world it would be nice if this block loader doesn't have to be changed if/when `semantic_text` fields store text in different way(s) and/or change suffix.

carlosdelest replied:
Text in `semantic_text` is not stored. It has to be retrieved from source.

ChrisHegarty replied:
👍

nik9000 reviewed (commented):
Makes sense to me. @not-napoleon might want to take a look because this is beginning to follow your lead.

ChrisHegarty reviewed (approved):
LGTM

ChrisHegarty committed:
sort

ChrisHegarty committed:
Merge branch 'main' into semantic_text_support

carlosdelest reviewed (approved):
`semantic_text` support, yay! 💯

ioanatia committed:
Attempt another fix for bwc tests

ioanatia commented:
There are some issues with loading this plugin for multi-cluster and mixed-versions when one of the cluster nodes in on 8.16.0:
```
       Caused by:
        java.lang.RuntimeException: Execution of elasticsearch-plugin failed with exit code 70
            at org.elasticsearch.test.cluster.local.AbstractLocalClusterFactory$Node.runToolScript(AbstractLocalClusterFactory.java:795)
            at org.elasticsearch.test.cluster.local.AbstractLocalClusterFactory$Node.installPlugins(AbstractLocalClusterFactory.java:633)
            at org.elasticsearch.test.cluster.local.AbstractLocalClusterFactory$Node.start(AbstractLocalClusterFactory.java:159)
            at org.elasticsearch.test.cluster.local.DefaultLocalClusterHandle.lambda$start$0(DefaultLocalClusterHandle.java:78)
            at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
            at java.base/java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:722)
            at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
            at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
            at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:759)
            at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
            at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
            at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
            at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
            at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
```

multi-node and single-node should continue to work just fine.
For now I conditionally disabled loading the plugin in multi-cluster and mixed-versions when one of the cluster nodes is on 8.x. Once we get a fix for the plugin, we can remove this check, but I don't have that much familiarity with the plugin and don't know how to fix it. I reached out to the team that owns it.

The checks `CsvTestsDataLoader` have also been changed - they no longer check for `SEMANTIC_TEXT_TYPE` ESQL capability, but rather check whether a dataset requires the test inference endpoint. This kind of makes more sense than to check for an ES|QL capability, we are not using ES|QL APIs to load the data, but the create index and bulk APIs.
If the test inference endpoint does not exist and a dataset requires it, we simply do not load the dataset.
In `EsqlSpecTestCase` I added another check for `shouldSkipTest` - in case `SEMANTIC_TEXT_TYPE` is required we also check that we can use the test inference service. This makes sense because if the test inference service we have no data with `semantic_text` fields.

ioanatia committed:
Spotless

ChrisHegarty commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

ChrisHegarty commented on x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java:
```diff
@@ -14,15 +14,21 @@
+
+        // The plugin cannot be loaded on mixed version clusters with versions on 8.x
+        if (oldVersion.onOrAfter("9.0.0")) {
+            cluster = cluster.plugin("inference-service-test");
```
For my one knowledge, why can the service not be loaded on ES 8.x ? Is this a short-term limitation or something else?

ioanatia replied:
It is a short-term limitation until we fix the plugin

ioanatia replied:
the only consequence for now is that we won't execute the test from `x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec` for mixed-version and multi-cluster

we do execute it for multi-node and single-node.
This only seems to happen when we try to set the plugin for clusters with 8.x nodes.

ChrisHegarty replied:
Ok, got it. This workaround for the temporary limitation seems reasonable. Is there an issue tracking the plugin issue?    

Also, just to confirm. Does the test plugin work ok on an 8.x only cluster - not mixed, just 8.x nodes. For example, if this change were to be backported to the Elasticsearch 8.x branch.

ioanatia replied:
We have another plugin `test-infererence-plugin` which is used to test retrievers for which we don't have any mixed version clusters or multi cluster tests.
Because of this I think it is fine we figure out how to enable these tests as a follow up and potentially find a solution we can use not just for ES|QL integration tests.

ioanatia committed:
Merge branch 'main' into semantic_text_support

ioanatia committed:
Fix merge

ioanatia committed:
Attempt another fix

ioanatia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

ioanatia committed:
Don't load the inference-service-test plugin for mixed versions/clusters

ioanatia commented:
@elasticmachine test this please

ioanatia committed:
Merge branch 'main' into semantic_text_support

ioanatia committed:
Merge branch 'main' into semantic_text_support

ioanatia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

fang-xing-esql commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-semantic_text.json:
```diff
@@ -0,0 +1,8 @@
+{
```
I wonder if we can add a few more fields into the mapping, like `KEYWORD`, `TEXT`, some numeric types etc? The mapping looks a bit simple to me. If `SEMANTIC_TEXT` will be in the same category as `KEYWORD` and `TEXT`, we might want to support some functions that apply to them as well(perhaps in future PRs). It will be handy to have some extra fields in the same mapping to validate how  processing commands and functions work with the new data type, like eval/where/keep/drop/sort/stats/rename/mv_expand etc.

ChrisHegarty replied:
done [e617196](https://github.com/elastic/elasticsearch/pull/113920/commits/e617196bfc36c52a17b3eb8bbc28c193924a4975).  More field types and functions can come later.

fang-xing-esql commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv:
```diff
@@ -0,0 +1,4 @@
+semantic_text_field:semantic_text
```
I wonder if `SEMANTIC_TEXT` makes sense for multi-valued field? It might be nice to have some multi-valued data to experiment with.

ioanatia replied:
Yes it absolutely does! I was not sure whether I should add a test for it here, but I did anyway for `MV_EXPAND`

fang-xing-esql commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec:
```diff
@@ -0,0 +1,10 @@
+simple
```
Can we add a few more queries with more processing commands here, like eval/where/keep/drop/sort/stats/rename/mv_expand etc? It seems like some of the processing commands might already work out of the box, and having more tests to show the coverage will help to clarify the scope of the PR. #110205 enabled date_nano on a few scalar functions, it seems like we are not there yet, but perhaps it works in a few processing commands already, it will be great if the tests can show more besides `SORT`.

ioanatia replied:
added tests for `eval/where/keep/drop/sort/stats/rename/mv_expand`
for functions and operators there is no support - I am doing that as a follow up here https://github.com/ioanatia/elasticsearch/compare/semantic_text_support...ioanatia:elasticsearch:semantic_text_as_string?expand=1
I took the dataset/mapping I had in my follow-up branch and added it here - the dataset has many fields that are not used in the tests yet, but will be used to test things like conversion functions when we add support for those.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java:
```diff
@@ -114,7 +114,7 @@ private static Object valueAt(DataType dataType, Block block, int offset, BytesR
             case INTEGER, COUNTER_INTEGER -> ((IntBlock) block).getInt(offset);
             case DOUBLE, COUNTER_DOUBLE -> ((DoubleBlock) block).getDouble(offset);
-            case KEYWORD, TEXT -> ((BytesRefBlock) block).getBytesRef(offset, scratch).utf8ToString();
+            case KEYWORD, SEMANTIC_TEXT, TEXT -> ((BytesRefBlock) block).getBytesRef(offset, scratch).utf8ToString();
```
Minor suggestion, swap `SEMANTIC_TEXT` and `TEXT`.

ioanatia replied:
No problem with changing this - but I am curious what's the reason? Right now they are in alphabetical order.
If someone takes a look at this code a year from now, they might appreciate having these in alphabetical order rather than having `SEMANTIC_TEXT` at the end which specifies that support was added for `SEMANTIC_TEXT` after `KEYWORD` and `TEXT` are available.

ChrisHegarty replied:
When I wrote this code originally, and many other similar switches in the compute engine, I deliberately did not follow a strict style consistency across the codebase, but rather preferred a logical and type grouping style. With a preference for more common and narrower types earlier in the cases.
I have no strong preference of ordering of this particular case label here.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/PositionToXContent.java:
```diff
@@ -89,7 +89,7 @@ protected XContentBuilder valueToXContent(XContentBuilder builder, ToXContent.Pa
                 }
             };
-            case KEYWORD, TEXT -> new PositionToXContent(block) {
+            case KEYWORD, SEMANTIC_TEXT, TEXT -> new PositionToXContent(block) {
```
Minor suggestion, swap `SEMANTIC_TEXT` and `TEXT`.

ChrisHegarty replied:
see https://github.com/elastic/elasticsearch/pull/113920#discussion_r1806171017

fang-xing-esql reviewed (commented):
I left the first round of comments, they are mainly about the scope of this PR. Adding a new data type in ES|QL is not trivial, it makes sense to break it down to multiple steps. If I understand it correctly(please correct me if I'm wrong), this PR enables `SEMANTIC_TEXT` in **some** processing commands, if there are more tests to show the coverage and scope that will be very appreciated.

I'll need more time to look into inference test service related code.

Thank you for labelling `test-release`, it takes a little longer to run, however it did help me catch some unexpected errors.

ioanatia committed:
Add more tests, address review comments

ioanatia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

fang-xing-esql reviewed (commented):
I left a few more comments. This is a subtask of supporting `SEMANTIC_TEXT` in `ES|QL`. I wonder if there is an umbrella issue created to keep track of all of the subtasks, similar to `DATE_NANOS`, so that we know where we are, and what to expect?

ChrisHegarty committed:
trivial

ChrisHegarty committed:
revert

ChrisHegarty committed:
Merge branch 'main' into semantic_text_support

ChrisHegarty committed:
post-merge fix block loader

ChrisHegarty committed:
post-merge fix compile

ChrisHegarty commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:
```diff
@@ -504,3 +504,52 @@ double nested declared in mapping:
+            properties:
+              semantic_text_field:
+                type: semantic_text
+                inference_id: my_inference_id
```
The compute engine is not actually doing any inference calls at this point, so the value here is for the most part irrelevant. However, it should likely be the same as that of what was configured by in the inference service (see `createInferenceEndpoint`), so `test_sparse_inference`.

ChrisHegarty replied:
There is no lookup of inference yet. That will come in a subsequent step (see the meta issue).

ChrisHegarty commented:
> I wonder if there is an umbrella issue created to keep track of all of the subtasks, similar to `DATE_NANOS`, so that we know where we are, and what to expect?

Good idea. Done #115103

ChrisHegarty commented:
FYI - I am removing the `test-release` label for now. It's adding too much latency and slowing iteration on this PR. It will be added before merge, at which time any failing tests it catches will be fixed.

ChrisHegarty committed:
add mixed version testing

ChrisHegarty committed:
whitespace

ChrisHegarty committed:
fix MultiClusterSpecIT

ChrisHegarty committed:
add more fields to mapping

ChrisHegarty commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:
```diff
@@ -207,6 +220,10 @@ protected static void checkCapabilities(RestClient client, TestFeatureService te
         }
     }
 
+    protected boolean supportsInferenceTestService() {
```
I fixed the registration of the inference plugin for mixed and multi-cluster tested.

ChrisHegarty committed:
Revert  mixed version testing

ChrisHegarty commented:
Ok, mixed node testing will be avoided until the following issue is resolved as a follow up. #115166.  (given that mixed mode will only be relevant when this PR is backported)

ChrisHegarty committed:
whitespace

ChrisHegarty commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

ChrisHegarty reviewed (approved):
This PR has had a number of review iterations. It's a great first step towards support for semantic search in ES|QL. LGTM 👍

ChrisHegarty commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:
```diff
@@ -504,3 +504,52 @@ double nested declared in mapping:
+                  "model_settings": {
+                    "task_type": "sparse_embedding"
+                  },
+                  "chunks": [{ "text": "be excellent to each other", "embeddings": { "a": 1,"b": 2 } }]
```
This is just how one helps the test infra translate to embeddings. It's mostly not interesting until the engine actually attempts the inference. Which will come in a later stage. (see the meta issue).

ChrisHegarty commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

ChrisHegarty commented:
@elasticmachine update branch

elasticmachine commented:
There are no new commits on the base branch.

ChrisHegarty commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

ioanatia committed:
Merge branch 'main' into semantic_text_support

ioanatia commented:
@elasticmachine test this please

ioanatia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into semantic_text_support

ioanatia commented:
Thank you folks, especially @fang-xing-esql and @ChrisHegarty for your help on this PR!
All the comment have been addressed, CI is green, the PR has a final approval from Chris after we addressed the comments, so it looks good to go 🚀

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 113920`
